File: src/main/java/work/lclpnet/ap2/game/fine_tuning/FineTuningInstance.java
package work.lclpnet.ap2.game.fine_tuning;

import net.minecraft.util.Identifier;
import net.minecraft.world.GameMode;
import org.json.JSONArray;
import work.lclpnet.ap2.api.game.MiniGameHandle;
import work.lclpnet.ap2.api.game.data.DataContainer;
import work.lclpnet.ap2.api.map.MapFacade;
import work.lclpnet.ap2.impl.game.BootstrapMapOptions;
import work.lclpnet.ap2.impl.game.DefaultGameInstance;
import work.lclpnet.ap2.impl.game.data.ScoreTimeDataContainer;
import work.lclpnet.kibu.scheduler.Ticks;

import java.util.Map;
import java.util.UUID;

public class FineTuningInstance extends DefaultGameInstance {

    static final int MELODY_COUNT = 3;
    static final int REPLAY_COOLDOWN = Ticks.seconds(5);
    private final ScoreTimeDataContainer data = new ScoreTimeDataContainer();
    private FineTuningSetup setup;
    private TuningPhase tuningPhase;
    private Map<UUID, FineTuningRoom> rooms;

    public FineTuningInstance(MiniGameHandle gameHandle) {
        super(gameHandle);

        setDefaultGameMode(GameMode.SURVIVAL);  // survival is needed to left-click note blocks
    }

    @Override
    protected DataContainer getData() {
        return data;
    }

    @Override
    protected void openMap() {
        MapFacade mapFacade = gameHandle.getMapFacade();
        Identifier gameId = gameHandle.getGameInfo().getId();

        mapFacade.openRandomMap(gameId, new BootstrapMapOptions((world, map) -> {
            setup = new FineTuningSetup(gameHandle, map, world);
            return setup.createRooms();
        }), this::onMapReady);
    }

    @Override
    protected void prepare() {
        JSONArray json = getMap().requireProperty("room-note-blocks");

        var noteBlockLocations = FineTuningSetup.readNoteBlockLocations(json, gameHandle.getLogger());
        setup.teleportParticipants(noteBlockLocations);

        rooms = setup.getRooms();
    }

    @Override
    protected void ready() {
        tuningPhase = new TuningPhase(gameHandle, rooms, data, this::startStagePhase);
        tuningPhase.init();
        tuningPhase.beginListen();
    }

    private void startStagePhase() {
        tuningPhase.unload();

        StagePhase stagePhase = new StagePhase(gameHandle, data, tuningPhase.getMelodies(), getMap(), getWorld(),
                winner -> winner.ifPresentOrElse(this::win, this::winNobody));

        stagePhase.beginStage();
    }
}

File: src/minigames/fine_tuning/src/main/kotlin/work/lclpnet/ap2/game/fine_tuning/FineTuningInstance.kt
package work.lclpnet.ap2.game.fine_tuning

import net.minecraft.server.level.ServerLevel
import org.json.JSONArray
import work.lclpnet.ap2.api.stats.Stat
import work.lclpnet.ap2.game.MiniGameHandle
import work.lclpnet.ap2.game.base.FFAGameInstance
import work.lclpnet.ap2.game.util.finaleCompatibleScoreContainer
import work.lclpnet.ap2.game.util.useAnnouncer
import work.lclpnet.ap2.game.util.useSurvivalMode
import work.lclpnet.ap2.impl.game.data.type.PlayerRef
import work.lclpnet.game.map.GameMap
import java.util.*

const val MELODY_COUNT = 2

val PitchChanges = Stat("pitch_changes", 0)
val Probes = Stat("probes", 0)
val Replays = Stat("replays", 0)
val MelodiesCompleted = Stat("melodies_completed", 0)
val CorrectNotes = Stat("correct_notes", 0)

class FineTuningInstance(
    gameHandle: MiniGameHandle,
    level: ServerLevel,
    map: GameMap,
    private val setup: FineTuningSetup,
) : FFAGameInstance(gameHandle, level, map) {

    override val data = finaleCompatibleScoreContainer(gameHandle, PlayerRef::create)
    private val stats = createStats(data, PitchChanges, Probes, Replays, MelodiesCompleted, CorrectNotes)
    private val announcer = useAnnouncer()
    private lateinit var tuningPhase: TuningPhase

    init {
        useSurvivalMode()
    }

    override fun prepare() {
        val json: JSONArray = map.requireProperty("room-note-blocks")
        val noteBlockLocations = FineTuningSetup.readNoteBlockLocations(json, gameHandle.logger)
        setup.teleportParticipants(noteBlockLocations)

        val rooms: Map<UUID, FineTuningRoom> = setup.rooms

        tuningPhase = TuningPhase(gameHandle, rooms, data, stats, ::startStagePhase, level, announcer)
        tuningPhase.init()
        tuningPhase.giveBooks()
    }

    override fun go() {
        tuningPhase.beginListen()
    }

    private fun startStagePhase() {
        tuningPhase.unload()

        val stagePhase = StagePhase(gameHandle, tuningPhase.records, map, level, winManager)
        stagePhase.beginStage()
    }
}
